keps/sig-apps/3715-elastic-indexed-job/README.md

```diff
@@ -0,0 +1,936 @@
+  - [ ] (R) Ensure GA e2e tests meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [x] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Currently `spec.completions` is an immutable in Job. This KEP proposes a new
+`CompletionMode` named `ElasticIndexed` that behaves similar to `Indexed` Job,
+but allows to mutate both `spec.completions` and `spec.parallelsim` if done together.
+
+In `ElasticIndexed`, `spec.completions` and `spec.parallelsim` are validated to always be equal. 
+Success, failure and completion semantics are exactly the same as `Indexed` Job.
+
+
+## Motivation
+
+There are cases where a batch workload requires an autoscaled indexed job with stable 
+DNS pod names, for example MPI Horovord, Ray, PyTorch etc. This is currently not possible 
+because `spec.completions`, which controls the range of indexes in `Indexed` job, is immutable. 
+While such workloads could be modeled as a StatefulSet, the Job API is a better fit because it
+offers batch-specific features such as allowing indexes to run to completion, and [pod and container failure handling](https://kubernetes.io/docs/concepts/workloads/controllers/job/#handling-pod-and-container-failures).
+
+In the above mentioned cases, there is no distinction between the concept of "how many completed 
+indexes do I need" (i.e., `spec.completions`) and the concept of "how many pods do I want running
+at a time" (i.e., `spec.parallelism`). All indexes are expected to start at the same time, but they
+can finish at different times.
+
+The above behavior can be modeled using an Indexed job with
+- `spec.completions` equal to `spec.parallelism`
+- mutable `spec.completions` and `spec.parallelsim` as long as they are mutated together.
+
+To represent this mode of operation for Job, this KEP introduces a new
```

Comment:
From a UX point of view I don't see any value from forcing user to always set both of these fields to the exact same value, rather than leaving `.spec.completions` set to `nil` and immutable and only modifying `.spec.parallelism` (as it's currently with non-index job). 
I totally understand @liggitt's concerns about breaking users' expectations about modifying the current IndexedJobs `.spec.completions` (and as usual he's probably right :wink: ) but I'm not sure that throwing the burden of setting both fields to user is a good option here. Especially that we are introducing entirely new CompletionMode.

Reply:
I would also have preferred if this case is possible by modifying a single field, but I don't see a major problem of having two to be consistent with the general case. Especially if the scale subresource idea doesn't have a good chance in the future.